libvmi/libvmi.h

```diff
@@ -572,6 +572,19 @@ typedef struct registers {
     };
 } registers_t;
 
+typedef struct hvm_cpu_xsave {
+    uint64_t xcr0;                 /* Updated by XSETBV */
+    struct {
+        struct { char x[512]; } fpu_sse;
+
+        struct hvm_hw_cpu_xsave_hd {
+            uint64_t xstate_bv;         /* Updated by XRSTOR */
+            uint64_t xcomp_bv;          /* Updated by XRSTOR{C,S} */
+            uint64_t reserved[6];
+        } xsave_hdr;                    /* The 64-byte header */
+    } save_area;
+} xen_cpu_xsave_info_t;
```

Comment:
This seems to be a 1:1 copy from a Xen header. I don't recommend copying these directly but rather define a LibVMI specific structure that only holds the fields you actually need and document what you need these fields for. FOr example, if you only need `xcr0` then instead of creating a whole new API you should just extend `xen_get_vcpureg_hvm` to return XCR0 if queried for.

Reply:
I need xcr0 and save_area from that xsave structure. The structure is not 1:1 with the xen one. I've dropped the 
 uint64_t xfeature_mask;        /* Ignored */
 uint64_t xcr0_accum;           /* Updated by XSETBV */

Reply:
The point of LibVMI is to abstract away the hypervisor specific bits and make the API applicable across-hypervisors (ideally). You are breaking that design here: you introduce a structure that is includes stuff like `xen` and `hvm`. The structure also have fields that are unexplained and their usecase is undocumented with LibVMI. What's `fpu_sse` used for? Nothing? Drop it. Why does it have `reserved` fields in there when it's not exposed through LibVMI as an ABI? So  this is a badly designed API right now.

Reply:
We use the SSE regs so fpu_sse is needed. I can drop the reserved filed but for the rest, we need that data. Is this ok or is this patch not usable?

Reply:
As I said, bleeding Xen specific stuff into the LibVMI public header is a no-go. If you can clean up the API, and especially if you could introduce an API that is more extendable going forward, that would be preferred. The LibVMI API should also be as architecture -neutral as possible. So give some thought on how to do that. Introducing new APIs for stuff that are pretty much just pass-through from other accessible libraries is IMHO pretty much pointless.

Reply:
That sounds like a hacky solution to me

Reply:
Yes, I agree but then what would be a good way to approach this or is the xsave a no go?

Reply:
Not in this current form, no. If you can come up it with a cleaner API then it would be considered.

Reply:
Hello Tamas,

In libbdvmi, we get the XSAVE area with the help of two helper functions, `getXSAVESize()`:

https://github.com/bitdefender/libbdvmi/blob/master/src/xendriver.cpp#L710

and `getXSAVEArea()`:

https://github.com/bitdefender/libbdvmi/blob/master/src/xendriver.cpp#L748

We need this in order to be able to work with the SSE registers. The most abstract way of doing this, AFAICT, is to simply fill up a user-provided (`void *`) buffer with the XSAVE area contents. If you look at "13.4 XSAVE AREA" in the Intel SDM, that's quite a lot of information in there, so A) it's not possible to decide on what is relevant and what is not, for the user, in all that, and B) if you look at how Xen and KVM look at that data, for example, you'll see that they're using different (small) parts of it.

We're fine passing this buffer along to our introspection engine to deal with it as it pleases, so for us the `void *` strategy is a perfectly fine use case. So we're not sure what you see as a better option for it, but if you'll let us know we'll try to fit the work into your design preferences.

Thanks.

Reply:
@razvan-cojocaru I don't have anything against XSAVE itself or having a way to get these required bits of information exposed through the LibVMI API. I already expressed my concerns with how this PR intends to add support to it to the public API - including names like `xen` and a structure that looks as-if it was an ABI when it's not and including garbage fields. If those get fixed, the PR will be considered. Adding a void* structure return here would just make the API even more incomprehensible so that route would be -1 from me.

Reply:
Fair enough, re-reading @aisaila's patch I think I know what you mean. :)
We'll re-shape the patch.